### PR TITLE
Transformation helpers

### PR DESCRIPTION
**Pull request by Karnaukhov-kh**

### CODE REVIEW

**Comment by Karnaukhov-kh** [comment]
Few changes comapred to original version:
- ```merge``` renamed to ```mergeObjects```. RxJs has 2 merges, third one exported from path that also includes 'rx' will confuse people.
- ```toArray``` renamed to ```dictionaryToArray```. Because method mainly targeting dictionaries.
- To handle edge cases some good chunk of checks added to helpers after tests were written.
- NoNullable removed to reduce noise. There are couple of issues related to this type. For example [this one](https://github.com/microsoft/TypeScript/issues/28374). If user passes null as some of arguments it will be handled inside of method anyway. And overall it feels redundant after tests were written because user still can do "null as any", null may appear over time but typescript will not detect it etc.

File: libs/state/docs/api/transformation-helpers/object/merge-objects.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ id: number,
+ type: string,
+ name: string
+}
+
+const cat = {id: 1, type: 'cat'};
+
+const catWithname = mergeObjects(cat, {name: 'Fluffy'});
+
+// anonymusCat will be:
```
Comment (kajetansw) [inline libs/state/docs/api/transformation-helpers/object/merge-objects.md:20]:
```suggestion
// catWithname will be:
```
Variable name doesn't match.

Comment (Karnaukhov-kh) [reply libs/state/docs/api/transformation-helpers/object/merge-objects.md:20]: Good catch, thank you. 👍

File: libs/state/docs/api/transformation-helpers/object/merge-objects.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+_Example_
+
+```TypeScript
+interface Creature {
+ id: number,
+ type: string,
+ name: string
+}
+
+const cat = {id: 1, type: 'cat'};
```
Comment (kajetansw) [inline libs/state/docs/api/transformation-helpers/object/merge-objects.md:16]:
```suggestion
const cat = {id: 1, type: 'cat', name: 'Leo'};
```
From what I've seen about the type `function mergeObjects<T extends object>(object: T, upd: Partial<T>): T` one field is missing from object above. Here the `upd` input has a field that `object` input doesn't.

File: libs/state/docs/api/transformation-helpers/object/merge-objects.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+Not mutating original object.
+
+_Example_
+
+```TypeScript
+interface Creature {
+ id: number,
+ type: string,
+ name: string
+}
```
Comment (Karnaukhov-kh) [inline libs/state/docs/api/transformation-helpers/object/merge-objects.md:14]:
```suggestion
```

File: libs/state/docs/api/transformation-helpers/object/merge-objects.md (hunk 2 above)
Comment (Karnaukhov-kh) [reply libs/state/docs/api/transformation-helpers/object/merge-objects.md:11]: Yes, nice. Also we can remove interface above because it is redundant.

File: libs/state/docs/api/transformation-helpers/object/merge-objects.md (hunk 1 above)
Comment (Karnaukhov-kh) [inline libs/state/docs/api/transformation-helpers/object/merge-objects.md]:
```suggestion
// catWithname will be:
```

**Comment by Karnaukhov-kh** [comment]
Doing battle testing on real project now. Found few edge cases when integrating helpers with connect() method and lazy state. So more updates will come today or tomorrow.

File: libs/state/spec/transformation-helpers/array/toDictionary.spec.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      expect(toDictionary('' as any, '')).toEqual({});
+      expect(toDictionary({} as any, '')).toEqual({});
+      expect(toDictionary(1 as any, '')).toEqual({});
+      expect(toDictionary(false as any, '')).toEqual({});
+    });
+
+    it('should return null or undefined when array is not provided', () => {
+      const arr: any[] = null as any;
+      const arr2: any[] = undefined as any;
+      expect( toDictionary(arr, '')).toEqual(undefined);
```
Comment (BioPhoton) [inline libs/state/spec/transformation-helpers/array/toDictionary.spec.ts:102]: shouldn't this return null?

File: libs/state/src/lib/transformation-helpers/array/insert.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/**
+ * @description
+ * Inserts one or multiple items to array T[].
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/array/insert.ts:3]:
```suggestion
 * Inserts one or multiple items to an array T[].
```

File: libs/state/src/lib/transformation-helpers/array/insert.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * Returns new updated array T[].
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/array/insert.ts:4]:
```suggestion
 * Returns a new instance of the updated array T[].
```

File: libs/state/src/lib/transformation-helpers/array/insert.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * Not mutating original array.
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/array/insert.ts:5]:
```suggestion
 * Returns a new instance of the updated array T[], and does not mutate the original one.
```

File: libs/state/src/lib/transformation-helpers/array/insert.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * @example
+ * // Inseting single value
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/array/insert.ts:8]:
```suggestion
 * // Inserting single value
```

File: libs/state/spec/transformation-helpers/array/toDictionary.spec.ts (hunk 4 above)
Comment (Karnaukhov-kh) [reply libs/state/spec/transformation-helpers/array/toDictionary.spec.ts:102]: Maybe yes. But select doing filter(slice => slice !== undefined). So idea was to return undefined even in case of null.

Overkill?

File: libs/state/src/lib/transformation-helpers/object/deleteProp.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import { isObjectGuard, isKeyOf } from '../../core/utils/typing';
+
+/**
+ * @description
+ * Accepts object of type T and key of type K extends keyof T.
+ * Removes property from an object and returns new updated object without specified property.
+ * If property not found returns copy of original object.
+ * Not mutating original object.
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/object/deleteProp.ts:8]:
```suggestion
 * Accepts an object of type T and key of type K extends keyof T.
 * Removes property from an object and returns a new updated object without specified property.
 * If property not found returns copy of the original object.
 * Not mutating original object.
```

File: libs/state/src/lib/transformation-helpers/object/dictionaryToArray.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import { isObjectGuard } from '../../core/utils/typing';
+
+/**
+ * @description
+ * Converts dictionary of type {[key: string]: T} to array T[].
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/object/dictionaryToArray.ts:5]:
```suggestion
 * Converts a dictionary of type {[key: string]: T} to array T[].
```

File: libs/state/src/lib/transformation-helpers/object/mergeObjects.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import { isObjectGuard } from '../../core/utils/typing';
+/**
+ * @description
+ * Merges object of type T with updates of type Partial T.
+ * Returns new object where updates overrides original values.
+ * Not mutating original object.
+ *
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/object/mergeObjects.ts:7]:
```suggestion
 * Merges an object of type T with updates of type Partial<T>.
 * Returns a new object where updates override original values while not mutating the original one.

```

File: libs/state/src/lib/transformation-helpers/object/mergeObjects.ts [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+ *  id: number,
+ *  type: string,
+ *  name: string
+ * }
+ *
+ * const cat = {id: 1, type: 'cat'};
+ *
+ * const catWithname = mergeObjects(cat, {name: 'Fluffy'});
+ *
+ * // anonymusCat will be:
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/object/mergeObjects.ts:19]:
```suggestion
 * // anonymousCat will be:
```

File: libs/state/src/lib/transformation-helpers/object/setProp.ts [hunk 13]
```diff
@@ -0,0 +1,39 @@
+import { isObjectGuard, isKeyOf } from '../../core/utils/typing';
+/**
+ * @description
+ * Accepts object of type T, key of type K extends keyof T, and value of type T[K].
+ * Sets the property and returns new updated object.
+ * Not mutating original object.
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/object/setProp.ts:6]:
```suggestion
 * Accepts an object of type T, key of type K extends keyof T, and value of type T[K].
 * Sets the property and returns a newly updated object while not mutating the original one 
```

File: libs/state/src/lib/transformation-helpers/object/toggle.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
+import { OnlyKeysOfSpecificType } from '../interfaces/only-keys-of-specific-type';
+import { isKeyOf, isObjectGuard } from '../../core/utils/typing';
+/**
+ * @description
+ * Toggles a boolean property in the object.
+ * Accepts object of type T and key value of which is boolean.
+ * Toggles the property and returns new object.
+ * Not mutating original object.
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/object/toggle.ts:8]:
```suggestion
 * Toggles a boolean property in the object.
 * Accepts object of type T and key value of which is boolean.
 * Toggles the property and returns a new object, while not mutating the original one.
```

File: libs/state/spec/transformation-helpers/array/toDictionary.spec.ts (hunk 4 above)
Comment (Karnaukhov-kh) [reply libs/state/spec/transformation-helpers/array/toDictionary.spec.ts]: Fixed

**Comment by Karnaukhov-kh** [comment]
All changes applied. Return values and tests are fixed.

**Comment by hoebbelsB** [comment]
@Karnaukhov-kh & @BioPhoton do we have eloborated enough about the different return values to merge this?

**Comment by Karnaukhov-kh** [comment]
Here's a summary of all edge cases that I can think of and return values for them.

**Array**

Insert (creatures is array of objects)

```
insert(null as any, creatures) > creatures
insert(creatures, null as any) > creatures
insert(null as any, null as any) > []
insert('' as any, creatures) > creatures
insert(1 as any, creatures) > creatures
insert({} as any, creatures) > creatures
insert(false as any, creatures) > creatures
```

update

```
// fully mirrors insert behavior
update(null as any, creatures) > creatures
update(creatures, null as any) > creatures
update(null as any, null as any) > []
update('' as any, creatures) > creatures
update(1 as any, creatures) > creatures
update({} as any, creatures) > creatures
update(false as any, creatures) > creatures
```

remove

```
remove(null as any, creatures) > null
remove(creatures, null as any) > creatures
remove(null as any, null as any) > null
insert('' as any, creatures) > []
insert(1 as any, creatures) > []
insert({} as any, creatures) > []
insert(false as any, creatures) > []
```

toDictionary

```
toDictionary([] as any, 'fakeKey') > {}
toDictionary(creatures, 'fake' as any) > {} //key not found
toDictionary(creatures, 'breeds' as any) > {} //key value is array
toDictionary(creatures, null as any) > {}
toDictionary('' as any, '') > {}
toDictionary({} as any, '') > {}
toDictionary(1 as any, '') > {}
toDictionary(false as any, '') > {}
toDictionary(null as any, '') > null
toDictionary(undefined as any, '') > undefined
```

**Object**

deleteProp

```
deleteProp(primitiveState, null as any) > primitiveState
deleteProp(null as any, null as any) > null
deleteProp(undefined as any, undefined as any) > undefined
deleteProp('' as any, 'fake') > {}
deleteProp(1 as any, 'fake') > {}
deleteProp(false as any, 'fake') > {}
deleteProp([], 'concat') > {} // not touching array
```

setProp

```
setProp('' as any, 'fake' as any, 42) > {}
setProp(null as any, 'fake', 42) > {}
setProp(null as any, null as any, 42) > {}
setProp([primitiveState], 'concat', () => []) > {} // not touching array
setProp(primitiveState, 'fake' as any, 42) > {...primitiveState, fake: 42}
setProp(primitiveState, null as any, 42) > primitiveState
```

toggle

```
toggle(primitiveState, null as any) > primitiveState
toggle(null as any, null as any) > {}
toggle(primitiveState, 'str' as any) > primitiveState
toggle(primitiveState, 'newBoolean' as any) > {...primitiveState, newBoolean: true}
```

mergeObjects

```
mergeObjects({}, primitiveState) > primitiveState
mergeObjects(null as any, primitiveState) > primitiveState
mergeObjects(primitiveState, null as any) > primitiveState
mergeObjects(null as any, null as any) > {}
mergeObjects(primitiveState, '' as any) > primitiveState
mergeObjects('' as any, primitiveState) > primitiveState
mergeObjects('' as any, '' as any) > primitiveState
mergeObjects(primitiveState, [primitiveState] as any) > primitiveState
mergeObjects([primitiveState] as any, primitiveState) > primitiveState
mergeObjects([primitiveState] as any, [primitiveState] as any) > {}
```

dictionaryToArray

```
dictionaryToArray({}) > []
dictionaryToArray(null as any) > null
dictionaryToArray(undefined as any) > undefined
dictionaryToArray('' as any) > []
dictionaryToArray(1 as any) > []
dictionaryToArray(false as any) > []
dictionaryToArray([1,2,3] as any) > [] // this one is questionable as for me.
```

**Comment by BioPhoton** [comment]
Are the return values in the edge cases all immutable?

**Comment by BioPhoton** [comment]
Thanks for the great work!

**Comment by Karnaukhov-kh** [comment]
@BioPhoton updated list of returning values with comment in the end (new or original value passed as input)

File: libs/state/src/lib/transformation-helpers/array/remove.ts [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+ * You can provide a custom comparison function that should return true if items match.
+ * If no comparison is provided, an equality check is used by default.
+Returns a new instance of the updated array T[], and does not mutate the original one.
+ *
+ * @example
+ * // Removing value without comparison function
+ *
+ * const items = [1,2,3,4,5];
+ *
+ * const updatedItems = insert(items, [1,2,3]);
```
Comment (BioPhoton) [inline libs/state/src/lib/transformation-helpers/array/remove.ts:15]: shouldn't this be about removing?

Comment (Karnaukhov-kh) [inline libs/state/src/lib/transformation-helpers/array/remove.ts:15]:
```suggestion
 * const updatedItems = remove(items, [1,2,3]);
```

File: libs/state/docs/api/transformation-helpers/array/remove.md [hunk 16]
```diff
@@ -0,0 +1 @@
...

##### typeof: T[]

### itemsOrItem

##### typeof: I[] | I

### compare

##### typeof: CompareFn&#60;T&#62;
```
Comment (Karnaukhov-kh) [inline libs/state/docs/api/transformation-helpers/array/remove.md:1]:
```suggestion
# remove

Removes one or multiple items from array T[].
You can provide a custom comparison function that should return true if items match.
If no comparison is provided, an equality check is used by default.
Returns new updated array T[].
Not mutating original array.

_Example_

```TypeScript
// Removing value without comparison function

const items = [1,2,3,4,5];

const updatedItems = remove(items, [1,2,3]);

// updatedItems will be: [4,5];
```

_Example_

```TypeScript
// Removing values with comparison function

const creatures = [{id: 1, type: 'cat'}, {id: 2, type: 'unicorn'}, {id: 3, type: 'kobold'}];

const nonExistingCreatures = [{id: 2, type: 'unicorn'}, {id: 3, type: 'kobold'}];

const realCreatures = remove(creatures, nonExistingCreatures, (a, b) => a.id === b.id);

// realCreatures will be: [{id: 1, type: 'cat'}];
```

## Signature

```TypeScript
function remove<T, I extends T>(array: T[], itemsOrItem: I[] | I, compare?: CompareFn<T>): T[]
```

## Parameters

### array

##### typeof: T[]

### itemsOrItem

##### typeof: I[] | I

### compare

##### typeof: CompareFn&#60;T&#62;

```

File: libs/state/src/lib/transformation-helpers/array/remove.ts (hunk 15 above)
Comment (Karnaukhov-kh) [reply libs/state/src/lib/transformation-helpers/array/remove.ts:15]: Whoops.

**Comment by Karnaukhov-kh** [comment]
Hi. Summary of changes from last commits.

## "mergeObjects" is now "patch"

It's just a better and more understandable name.

## Concepts are coming together

_This is how things are now but can be changed if we decide so._

- All helpers that should extend source now will return updates (if provided) in cases when the source is not matching expected type (helper expected object but got a string for example) or is null/undefined.
``` 
Array: insert, update, 
Object: patch, setProp, toggle
```
- All helpers that should shrink the source will return empty object/array in cases when the source is not matching the needed type. In the case of null/undefined, we will return null/undefined because initializing source with a helper that should shrink it feels strange.
```
Array: remove
Object: deleteProp
```
- All helpers that perform a conversion from one type to another will return the empty target type in the case when the source not matching the expected type (expected array, got string). In the case of null/undefined, we will return null/undefined.
```
Array: toArray
Object: toDictionary
```

## Changed behaviors

### array/update

In case of the empty initial array will return provided updates

```
update([], [{id: 1, type: 'cat'}], (a, b) => a.id === b.id) // [{id: 1, type: 'cat'}]
```
Previously were returning an empty array.

**Reason:** to align with "insert" that returns updates as new array if original array not provided/empty/not array.


### object/setProp & object/toggle

In case when the initial object is not an object or not provided (undefined/null) will return a new object with the provided property.

```
setProp(undefined as any, 'num', 42) // {num: 42}. Same for anything that is not an object
toggle(undefined as any, 'bol') // {bol: true}. Same for anything that is not an object
```

**Reason:** to align better with "patch" that returns an object with updates when the original object not provided/empty/not an object. 

## Docs changes

Docs will now say "shallow copy" instead of "new instance". Since "new instance" can be easily misinterpreted and in reality, we are returning shallow copies.

File: libs/state/src/lib/index.ts [hunk 17]
```diff
@@ -3,6 +3,7 @@ export {
   select,
   stateful,
   distinctUntilSomeChanged,
-  selectSlice,
+  selectSlice
 } from './rxjs/operators';
 export { KeyCompareMap, CompareFn } from './rxjs/interfaces';
+export * from './transformation-helpers/index';
```
Comment (hoebbelsB) [inline libs/state/src/lib/index.ts:9]: can we specifiy the exports please?

File: libs/state/src/lib/transformation-helpers/array/remove.ts [hunk 18]
```diff
@@ -0,0 +1,106 @@
+import { isDefined } from '../../core/utils/typing';
+import { valuesComparer } from '../_internals/valuesComparer.util';
+import { ComparableData } from '../interfaces/comparable-data-type';
+
+/**
+ * @description
+ * Removes one or multiple items from an array T[].
+ * For comparison you can provide key, array of keys or a custom comparison function that should return true if items match.
```
Comment (hoebbelsB) [inline libs/state/src/lib/transformation-helpers/array/remove.ts:8]:
```suggestion
 * For comparison you can provide a key, an array of keys or a custom comparison function that should return true if items match.
```

File: libs/state/src/lib/transformation-helpers/index.ts [hunk 19]
```diff
@@ -0,0 +1,10 @@
+export * from './array/insert';
+export * from './array/remove';
+export * from './array/update';
+export * from './array/toDictionary';
+export * from './object/deleteProp';
+export * from './object/patch';
+export * from './object/setProp';
+export * from './object/dictionaryToArray';
+export * from './object/toggle';
+export * from './interfaces/only-keys-of-specific-type';
```
Comment (hoebbelsB) [inline libs/state/src/lib/transformation-helpers/index.ts:10]: create barrel files and export those

**Comment by Karnaukhov-kh** [comment]
Will close this one since we recreated and merged it as a new one from repo branch. #169
